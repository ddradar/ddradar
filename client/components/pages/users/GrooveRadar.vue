<template>
  <reactive-radar :chart-data="chartData" :chart-options="chartOptions" />
</template>

<script lang="ts">
import type { Song } from '@ddradar/core'
import type { ChartData, ChartTooltipItem } from 'chart.js'
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import ReactiveRadar from '~/components/shared/ReactiveRadar'

@Component({ components: { ReactiveRadar } })
export default class GrooveRadarComponent extends Vue {
  @Prop({ type: Object, default: null })
  readonly chart!: Song.GrooveRadar | null

  get chartOptions() {
    return {
      legend: { display: false },
      responsive: true,
      scale: { ticks: { beginAtZero: true, max: 200, min: 0, stepSize: 20 } },
      tooltips: { enabled: true, callbacks: { label: this.renderLabel } },
    }
  }

  get chartData(): ChartData {
    return {
      labels: ['STREAM', 'CHAOS', 'FREEZE', 'AIR', 'VOLTAGE'],
      datasets: [
        {
          label: 'data',
          backgroundColor: 'rgba(0, 255, 255, 0.2)',
          borderColor: 'rgba(0, 192, 192, 0.5)',
          data: this.chart
            ? [
                this.chart.stream,
                this.chart.chaos,
                this.chart.freeze,
                this.chart.air,
                this.chart.voltage,
              ]
            : [],
        },
      ],
    }
  }

  renderLabel(
    { index }: Pick<ChartTooltipItem, 'index'>,
    { labels }: Pick<ChartData, 'labels'>
  ) {
    if (index === undefined || labels === undefined) return ''
    const dataSet = labels[index]
    return Array.isArray(dataSet) ? dataSet[0] : dataSet
  }
}
</script>
