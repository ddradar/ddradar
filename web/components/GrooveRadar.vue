<template>
  <Radar :data="chartData" :options="chartOptions" />
</template>

<script lang="ts" setup>
import type { ScoreSchema } from '@ddradar/core'
import type { ChartData, ChartOptions } from 'chart.js'
import { Radar } from 'vue-chartjs'

type GrooveRadar = Exclude<ScoreSchema['radar'], undefined | null>
const props = withDefaults(
  defineProps<{ radar: GrooveRadar; color?: string; ticks?: boolean }>(),
  {
    color: 'rgba(0, 192, 192, 0.5)',
    ticks: false,
  }
)
const colorMode = useColorMode()
const isDark = computed(() => colorMode.value === 'dark')

const chartData = computed<ChartData<'radar'>>(() => ({
  labels: ['STREAM', 'CHAOS', 'FREEZE', 'AIR', 'VOLTAGE'],
  datasets: [
    {
      label: '',
      backgroundColor: props.color,
      borderColor: props.color,
      data: [
        props.radar.stream,
        props.radar.chaos,
        props.radar.freeze,
        props.radar.air,
        props.radar.voltage,
      ],
    },
  ],
}))
const chartOptions = computed<ChartOptions<'radar'>>(() => ({
  responsive: true,
  plugins: { legend: { display: false } },
  scales: {
    r: {
      beginAtZero: true,
      max: 200,
      min: 0,
      ticks: {
        display: props.ticks,
        stepSize: 20,
        color: isDark.value ? 'rgba(255,255,255,0.3)' : undefined,
        backdropColor: 'rgba(0,0,0,0)',
      },
      grid: { color: isDark.value ? 'rgba(255,255,255,0.3)' : undefined },
      angleLines: { color: isDark.value ? 'rgba(255,255,255,0.3)' : undefined },
    },
  },
}))
</script>
