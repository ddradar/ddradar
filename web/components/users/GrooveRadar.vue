<template>
  <Radar :data="chartData" :options="chartOptions" />
</template>

<script lang="ts" setup>
import type { Database } from '@ddradar/core'
import type { ChartData, ChartOptions } from 'chart.js'
import { Radar } from 'vue-chartjs'

const props = defineProps<{ radar: Database.GrooveRadar }>()

const chartData: ChartData<'radar'> = {
  labels: ['STREAM', 'CHAOS', 'FREEZE', 'AIR', 'VOLTAGE'],
  datasets: [
    {
      label: '',
      backgroundColor: 'rgba(0, 192, 192, 0.5)',
      borderColor: 'rgba(0, 192, 192, 0.5)',
      data: [
        props.radar.stream,
        props.radar.chaos,
        props.radar.freeze,
        props.radar.air,
        props.radar.voltage,
      ],
    },
  ],
}
const chartOptions: ChartOptions<'radar'> = {
  responsive: true,
  plugins: { legend: { display: false } },
  scales: {
    r: { beginAtZero: true, max: 200, min: 0, ticks: { stepSize: 20 } },
  },
}
</script>
