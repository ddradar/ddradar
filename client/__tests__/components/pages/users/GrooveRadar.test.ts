import { mount, shallowMount, Wrapper } from '@vue/test-utils'

import GrooveRadar from '~/components/pages/users/GrooveRadar.vue'

describe('/components/pages/users/GrooveRadar.vue', () => {
  describe('snapshot test', () => {
    test('renders correctly', async () => {
      const chart = {
        stream: 120,
        voltage: 100,
        air: 266,
        freeze: 200,
        chaos: 140,
      }
      const wrapper = mount(GrooveRadar, { propsData: { chart } })
      await wrapper.vm.$nextTick()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('renderLabel', () => {
    let wrapper: Wrapper<GrooveRadar>
    beforeEach(() => (wrapper = shallowMount(GrooveRadar)))

    test.each([
      [{}, {}],
      [{ index: 0 }, {}],
      [{}, { labels: ['foo'] }],
      [{}, { labels: [['foo']] }],
    ])('(%p, %p) returns ""', (toolTip, data) => {
      // @ts-ignore
      expect(wrapper.vm.renderLabel(toolTip, data)).toBe('')
    })
    test.each([
      [0, ['foo'], 'foo'],
      [0, [['foo']], 'foo'],
      [1, ['foo', 'bar'], 'bar'],
      [1, [['foo', 'bar'], ['baz']], 'baz'],
    ])(
      '({ index: %i }, { labels: %p }) returns "%s"',
      (index, labels, expected) => {
        // @ts-ignore
        expect(wrapper.vm.renderLabel({ index }, { labels })).toBe(expected)
      }
    )
  })
})
