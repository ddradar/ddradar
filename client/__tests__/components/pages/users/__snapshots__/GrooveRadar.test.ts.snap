// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/pages/users/GrooveRadar.vue snapshot test renders correctly 1`] = `<div class=""><canvas id="radar-chart" width="0" height="0" style="display: block;" class="chartjs-render-monitor"></canvas></div>`;
